src/ir.cpp

```diff
@@ -15669,6 +15738,323 @@ static TypeTableEntry *ir_analyze_instruction_offset_of(IrAnalyze *ira,
+                    payload->data.x_struct.parent.data.p_union.union_val = parent;
+                }
+                else if (parent->type->id == TypeTableEntryIdUnion)
+                {
+                    payload->data.x_struct.parent.id = ConstParentIdUnion;
+                    payload->data.x_struct.parent.data.p_union.union_val = parent;
+                }
+                else
+                {
+                    payload->data.x_struct.parent.id = ConstParentIdNone;
+                }
+                // child: ?&TypeInfo
+                ConstExprValue *type_info_type = get_builtin_value(ira->codegen, "TypeInfo");
+                assert(type_info_type->type->id == TypeTableEntryIdMetaType);
+
+                TypeTableEntry *type_info_ptr_type = get_pointer_to_type(ira->codegen, type_info_type->data.x_type, false);
+
+                fields[0].special = ConstValSpecialStatic;
+                fields[0].type = get_maybe_type(ira->codegen, type_info_ptr_type);
+
+                if (type_entry->data.promise.result_type == nullptr)
+                {
+                    fields[0].data.x_maybe = nullptr;
+                }
+                else
+                {
+                    ConstExprValue *maybe_value = create_const_vals(1);
+                    maybe_value->special = ConstValSpecialStatic;
+                    maybe_value->type = type_info_ptr_type;
+
+                    maybe_value->data.x_ptr.special = ConstPtrSpecialRef;
+                    maybe_value->data.x_ptr.mut = ConstPtrMutComptimeVar;
+
+                    ConstExprValue *union_val = create_const_vals(1);
+                    union_val->special = ConstValSpecialStatic;
+                    union_val->type = type_info_type->data.x_type;
+                    bigint_init_unsigned(&union_val->data.x_union.tag, type_id_index(type_entry->data.promise.result_type->id));
+                    union_val->data.x_union.payload = ir_make_type_info_value(ira, union_val, type_entry->data.promise.result_type);
+
+                    maybe_value->data.x_ptr.data.ref.pointee = union_val;
+                    fields[0].data.x_maybe = maybe_value;
+                }
+                return payload;
+            }
+        default:
+            zig_unreachable();
+    }
+}
+
+static TypeTableEntry *ir_analyze_instruction_type_info(IrAnalyze *ira,
+        IrInstructionTypeInfo *instruction)
+{
+    IrInstruction *type_value = instruction->type_value->other;
+    TypeTableEntry *type_entry = ir_resolve_type(ira, type_value);
+    if (type_is_invalid(type_entry))
+        return ira->codegen->builtin_types.entry_invalid;
+
+    ConstExprValue *var_value = get_builtin_value(ira->codegen, "TypeInfo");
+    assert(var_value->type->id == TypeTableEntryIdMetaType);
+    TypeTableEntry *result_type = var_value->data.x_type;
+
+    // TODO: We need to return a const pointer to the typeinfo, not the typeinfo by value.
```

Comment:
why is that?

Reply:
@andrewrk  

Well, tbh we don't "need" to, I just think it's cleaner if the compiler gives us a const ref to the TypeInfo instance.  

The user can then copy it if they need to.
It's probably just fine to return it by value too though, I don't know why it struck me as necessary in the first place :P